spec/index.bs

```diff
@@ -1233,29 +1222,25 @@ The {{HTMLScriptElement/src}} setter steps are:
     `script`.</ins>
 1.  <ins>Set [=this=]'s [=src=] content attribute to |value|.</ins>
 
-#### Setting slot values from parser #### {#setting-slot-values-from-parser}
+#### Script children changed steps #### {#script-children-changed-steps}
 
-This document modifies the HTML parser to set the [=script text=] value when the script is created.
+This document modifies the [=children changed steps=] for {{HTMLScriptElement}} as follows:
 
-Modify the [=The text insertion mode=] algorithm as follows:
+1.  Set [=this=]'s [=HTMLScriptElement/is trusted=] to false.
 
-<dl class="switch">
-    <dt id="scriptEndTag">An end tag whose tag name is "script"</dt>
-    <dd>
-        <p>...</p>
+1.  If [=this=]'s [=HTMLScriptElement/changed by trusted sink=] is true, set [=this=]'s [=HTMLScriptElement/is trusted=] to true.
 
-        <ins><p>Set <var>script</var>'s [=script text=] value to its [=child text content=].</p></ins>
+1.  Set [=this=]'s [=HTMLScriptElement/changed by trusted sink=] to false.
 
-        <p>If the <span>active speculative HTML parser</span> is null, then <span>prepare the script
-        element</span> <var>script</var>. This might cause some script to execute, which might cause
-        <span data-x="dom-document-write">new characters to be inserted into the tokenizer</span>, and
-        might cause the tokenizer to output more tokens, resulting in a [=reentrant invocation of the parser=].</p>
+Note: This relies on the children changed steps never being called by the parser.
```

Comment:
How can we assume this? I think these steps are called indirectly by the parser, in the usual insertion path, no? And it seems that browsers expect this too: https://source.chromium.org/chromium/chromium/src/+/main:third_party/blink/renderer/core/html/html_script_element.cc;l=82;drc=d81683e439e0b80cac98e017d94449d59f48fa0d.

Reply:
Apologies for the slow response I've been OoO I didn't neccesarily expect this PR to be merged until someone with more knowledge of the status quo had reviewed, apologies for not making that clearer.

So from all I could work out, the specification children changed steps are only called from APIs and NOT the parser. 
If that's not true then this will need reverting. 
The ChildrenChanged method implemented in Chromium and WebKit are called by the parser but that's because they aren't a 100% match for the spec.

Fwiw I did a quick prototype in Chromium and it seemed to work as expected (it was a few months back at this point so can't fully remember). I've also got a draft PR to update WebKit to use this new model too https://github.com/WebKit/WebKit/pull/30649.

Reply:
> Apologies for the slow response I've been OoO

Hah, not a problem, my delay since Anne pinged be about this was far worse.

> I didn't neccesarily expect this PR to be merged until someone with more knowledge of the status quo had reviewed, apologies for not making that clearer.

Do you think we should revert it?

> So from all I could work out, the specification children changed steps are only called from APIs and NOT the parser.
If that's not true then this will need reverting.

Hmm, spec-wise, I think the children changed steps are indeed called from the parser. At the very least, the HTML's adoption agency algorithm [creates an element](https://html.spec.whatwg.org/C#parsing-main-inbody:create-an-element-for-the-token) and then immediately appends it to the DOM, which goes through the https://dom.spec.whatwg.org/#concept-node-insert path, which itself [calls the children changed steps](https://dom.spec.whatwg.org/#ref-for-concept-node-children-changed-ext).

I am really not familiar with the HTML parser, so I can't find any other paths from parser -> DOM insertion hook (which calls children changed steps), but I'm pretty sure they exist. For example, when a script is run immediately after insertion to the DOM, it does so via the post-connection steps: https://html.spec.whatwg.org/C#script-processing-model:html-element-post-connection-steps. And the only thing that calls these steps [is the DOM insertion algorithm](https://dom.spec.whatwg.org/#ref-for-concept-node-post-connection-ext%E2%91%A5)—the same thing that calls the children changed steps—so it must be invoked during the parser. Same goes for iframes, which initialize in the same way.